torchtune/datasets/_stack_exchange_paired.py

```diff
@@ -0,0 +1,76 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Dict, Optional
+
+from torchtune.data import (
+    ChosenRejectedToMessages,
+    PromptTemplate,
+    QuestionAnswerTemplate,
+)
+from torchtune.datasets._preference import PreferenceDataset
+from torchtune.modules.tokenizers import ModelTokenizer
+
+
+def stack_exchange_paired_dataset(
+    tokenizer: ModelTokenizer,
+    *,
+    source: str = "lvwerra/stack-exchange-paired",
+    column_map: Optional[Dict[str, str]] = None,
+    prompt_template: Optional[PromptTemplate] = QuestionAnswerTemplate(),
+    train_on_input: bool = False,
+    split: str = "train",
+) -> PreferenceDataset:
+    """
+    Family of preference datasets similar to `StackExchangePaired data
+    <https://huggingface.co/datasets/lvwerra/stack-exchange-paired>`_.
+
+    Args:
+        tokenizer (ModelTokenizer): Tokenizer used by the model that implements the ``tokenize_messages`` method.
+        source (str): path to dataset repository on Hugging Face. For local datasets,
+            define source as the data file type (e.g. "json", "csv", "text") and pass
+            in the filepath in ``data_files``. See Hugging Face's ``load_dataset``
+            (https://huggingface.co/docs/datasets/en/package_reference/loading_methods#datasets.load_dataset.path)
+            for more details. Default is ``liweili/c4_200m``.
+        column_map (Optional[Dict[str, str]]): a mapping from the expected columns in the prompt template
+            to the new column names in the dataset. If None, assume these are identical.
+        prompt_template (Optional[PromptTemplate]): optional template used to format the prompt. Default
+            is :class:`~torchtune.data.QuestionAnswerTemplate`.
+        train_on_input (bool): Whether the model is trained on the prompt or not. Default is False.
+        split (str): ``split`` argument for ``datasets.load_dataset``. You can use this argument to load a subset
+            of a given split, e.g. ``split="train[:10%]"``. Default is "train".
+
+    Returns:
+        PreferenceDataset: The preference dataset built from source paired data.
+    """
+
+    column_map = column_map or {
+        "prompt": "question",
+        "chosen": "response_j",
+        "rejected": "response_k",
+    }
+
+    def _filter(x):
+        return (
+            len(x[column_map["prompt"]]) + len(x[column_map["chosen"]])
+            <= tokenizer.max_seq_len
+            and len(x[column_map["prompt"]]) + len(x[column_map["rejected"]])
+            <= tokenizer.max_seq_len
+        )
```

Comment:
I know this was in the original implementation, but is this really the correct usage? Like we are truncating a string to max seq len here, not the tokens? (Unless I'm missing something obvious)

Reply:
yeah, see here for the original useage

https://github.com/eric-mitchell/direct-preference-optimization/blob/f8b8c0f49dc92a430bae41585f9d467d3618fe2f/preference_datasets.py#L241

Reply:
This is something that's been on the back of my mind that has bothered me a little - this is specific to:

- the dataset format used in the original DPO experiments
- the use-case for the original DPO PR

There are many preference datasets which won't have the "prompt" column. I'd personally rather just see this removed as the default.

Reply:
Wait but this pointer is happening after tokenization, right? Aren't we doing this filtering prior to tokenization? In that case we're talking about len(str) here not len(List[int])

Reply:
ooh good point. I vote to just remove.

Reply:
Sorry, yeah, I'm getting muddled up here. I'd PREFER to remove.

What's the workflow for a user wanting to filter on the tokenized chosen/rejected? Define their own class?

Reply:
> to filter on the tokenized chosen/rejected

dont think this is a real use case, because what can you filter if everything is in integer ids? truncating past max seq len is already handled by the tokenizer

Reply:
Yeah okay, I guess with our default preference dataset we already have the chosen/rejected and they're just being tokenized - you don't to construct them and truncate correctly.